include/swift/Localization/LocalizationFormat.h

```diff
@@ -173,16 +175,28 @@ class LocalizationProducer {
   virtual llvm::StringRef getMessageOr(swift::DiagID id,
                                        llvm::StringRef defaultMessage);
 
+  /// \returns a `SerializedLocalizationProducer` pointer if the serialized
+  /// diagnostics file available, otherwise returns a `YAMLLocalizationProducer`
+  /// if the `YAML` file is available. If both files aren't available returns a
+  /// `nullptr`.
+  static std::unique_ptr<LocalizationProducer> producerFor(llvm::StringRef locale, llvm::StringRef path, bool printDiagnosticNames);
+
   virtual ~LocalizationProducer() {}
 
 protected:
+  LocalizationProducerState state;
+
+  /// Used to lazily initialize `LocalizationProducer`
+  virtual void initialize() = 0;
```

Comment:
I suggest you add a `virtual bool initializeImpl() = 0;` (return indicates whether initialization succeeded or failed) and change this one to:

```
virtual void initialize() final {
   if (initializeImpl())
    state = Initialized;
   else
    state = FailedInitialization;
}
```

This way `LocalizationProducer` controls initialization done by sub-classes and `state` flag is protected.

Reply:
I think a better way to do that would be to add a check in `initialize` to check whether the localization producer has been initialized. I.e. 
```
virtual void initialize() final {
  if (state == NotInitialized && initializeImpl())
    state = Initialized;
  else
    state = FailedInitialization;
}
```
That way we won't proceed to `initializeImpl()` unless we are sure that the localization producer hasn't been initialized before. 

By making that change, I think we can simplify the checking statement in `LocalizationProducer::getMessageOr` to:
```
LocalizationProducer::getMessageOr(swift::DiagID id,
                                   llvm::StringRef defaultMessage) {
  initialize();
  if (state == FailedInitialization) {
    return defaultMessage;
  }
...
...
```
and in `YAMLLocalizationProducer::forEachAvailable` it will also be:
```
void YAMLLocalizationProducer::forEachAvailable(
    llvm::function_ref<void(swift::DiagID, llvm::StringRef)> callback) {
  initialize();
...
...
...
```

Reply:
Sure, rename initialize to initializeIfNeeded then.

Reply:
Actually renaming that allows us to switch back to `initialize()` for implementation specific one, also:

```
if (state != NotInitialized)
  return;

if (initializeImpl())
  ...
else
  ...
```

Is better because otherwise `state` would be constantly reset for no reason.